src/string.rs

```diff
@@ -0,0 +1,194 @@
+pub type StringManyMatcher =
+    ManyMatcher<StringPattern, StringIndexKey, StringPredicate, StringIndexingScheme>;
+
+/// Simple indexing schemes for strings.
+///
+/// As strings form a chain of characters, the indexing scheme is very simple:
+/// just the integer index given by the position in the string.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub struct StringIndexingScheme;
+
+impl IndexingScheme<String, StringPosition> for StringIndexingScheme {
+    type Key = StringIndexKey;
+
+    fn valid_bindings(
+        &self,
+        key: &Self::Key,
+        known_bindings: &impl IndexMap<Self::Key, StringPosition>,
+        data: &String,
+    ) -> BindingOptions<Self::Key, StringPosition> {
+        if key == &Self::Key::root() {
+            // For the root binding, any string position is valid
+            BindingOptions::ValidBindings((0..data.len()).map_into().collect())
+        } else {
+            let Some(StringPosition(root_pos)) = known_bindings.get(&Self::Key::root()) else {
+                return BindingOptions::MissingIndexKeys(vec![Self::Key::root()]);
+            };
+
+            let &StringIndexKey(offset) = key;
+
+            BindingOptions::ValidBindings(vec![(root_pos + offset).into()])
+        }
+    }
+}
+
+/// A position in a string.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, From, Into)]
+pub struct StringPosition(usize);
+
+/// An index key, obtained from the position index of the character in the pattern.
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, From, Into)]
+pub struct StringIndexKey(usize);
+
+impl StringIndexKey {
+    fn root() -> Self {
+        Self(0)
+    }
+}
+
+impl Debug for StringIndexKey {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "char@{}", self.0)
+    }
+}
+
+#[cfg(test)]
+pub(super) mod tests {
+    use crate::HashMap;
+
+    use rstest::rstest;
+
+    use self::pattern::StringPattern;
+    use crate::{ManyMatcher, PatternMatch, PortMatcher};
+
+    use super::*;
+
+    fn non_det_matcher(patterns: Vec<StringPattern>) -> StringManyMatcher {
+        ManyMatcher::from_patterns_with_det_heuristic(patterns, |_| false)
+    }
+
+    fn default_matcher(patterns: Vec<StringPattern>) -> StringManyMatcher {
+        ManyMatcher::from_patterns(patterns)
+    }
+
+    fn det_matcher(patterns: Vec<StringPattern>) -> StringManyMatcher {
+        ManyMatcher::from_patterns_with_det_heuristic(patterns, |_| true)
+    }
+
+    pub(super) const MATCHER_FACTORIES: &[fn(Vec<StringPattern>) -> StringManyMatcher] =
+        &[non_det_matcher, default_matcher, det_matcher];
+
+    /// Comparing two lists of pattern matches.
+    ///
+    /// We cannot use equality for two reasons:
+    ///  - we ignore the ordering of the matches
+    ///  - not all keys may be bound depending on the order things were matched
+    ///
+    /// The simplest way to identify a match is with the pair (PatternID, start_pos).
```

Comment:
Is that only true for strings or does the same idea apply to matches on other kinds of data? (I.e., does this belong in `string.rs`? You'd need some parametrization of course)

Reply:
This was removed in the modified code, but to answer your question: it is a property of strings specifically. For general graphs for example, it is impossible to recover the entire subgraph isomorphism from just the image of one node.